File: app/src/main/java/com/marked/pixsee/signup/SignUpActivity.java
package com.marked.pixsee.signup;


import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.IntentCompat;
import android.support.v7.app.AppCompatActivity;
import android.widget.Toast;

import com.marked.pixsee.Pixsee;
import com.marked.pixsee.R;
import com.marked.pixsee.injection.modules.ActivityModule;
import com.marked.pixsee.main.MainActivity;
import com.marked.pixsee.signup.di.DaggerSignupComponent;
import com.marked.pixsee.signup.di.SignupModule;

import javax.inject.Inject;

public class SignUpActivity
		extends AppCompatActivity
		implements SignUpNameFragment.SignUpNameFragmentInteraction,
		SignUpEmailFragment.SignUpEmailFragmentInteraction,
		SignUpPassFragment.SignUpPassFragmentInteraction, SignUpContract.View {
	private FragmentManager mFragmentManager;
	private ProgressDialog mProgressDialog;

	@Inject
	SignUpContract.Presenter mPresenter;

	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_sign_up);
		mFragmentManager = getSupportFragmentManager();
		mProgressDialog = new ProgressDialog(this);
		DaggerSignupComponent.builder()
				.appComponent(((Pixsee) getApplication()).getAppComponent())
				.activityModule(new ActivityModule(this))
				.signupModule(new SignupModule(this))
				.build().inject(this);
	}

	@Override
	protected void onStart() {
		super.onStart();
		mPresenter.start();
	}

	@Override
	protected void onDestroy() {
		mProgressDialog.dismiss();
		super.onDestroy();
	}

	@Override
	public void onSaveName(@NonNull String name) {
		mPresenter.onSaveName(name);
	}

	@Override
	public void onSaveEmail(@NonNull String email) {
		mPresenter.onSaveEmail(email);
	}

	@Override
	public void showDialog(String title, String message) {
		mProgressDialog.setTitle(title);
		mProgressDialog.setMessage(message);
		mProgressDialog.setIndeterminate(true);
		mProgressDialog.show();
	}

	@Override
	public void signupComplete(String name, String email, String username, String password) {
	}

	@Override
	public void showMainScreen() {
		Intent toStart = new Intent(this, MainActivity.class);
		toStart.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
		toStart.addFlags(IntentCompat.FLAG_ACTIVITY_CLEAR_TASK);
		startActivity(toStart);
	}

	@Override
	public void onSavePassword(@NonNull String password) {
		mPresenter.onSavePassword(password);
	}

	@Override
	public void onSaveUsername(String username) {
		mPresenter.onSaveUsername(username);
	}

	@Override
	public void showSignupStepPassword() {
		mFragmentManager.beginTransaction()
				.add(R.id.fragmentContainer, SignUpPassFragment.newInstance())
				.addToBackStack("signupPass")
				.commit();
	}

	@Override
	public void showSignupStepName() {
		mFragmentManager.beginTransaction()
				.add(R.id.fragmentContainer, SignUpNameFragment.newInstance())
				.commit();
	}

	@Override
	public void showSignupStepEmail(String name) {
		mFragmentManager.beginTransaction()
				.add(R.id.fragmentContainer, SignUpEmailFragment.newInstance(name))
				.addToBackStack("signupEmail")
				.commit();
	}

	@Override
	public void showToast(String message) {
		Toast.makeText(SignUpActivity.this, message, Toast.LENGTH_SHORT).show();
	}

	@Override
	public void hideDialog() {
		mProgressDialog.hide();
	}

	@Override
	public void setPresenter(SignUpContract.Presenter presenter) {
		mPresenter = presenter;
	}
}

File: app/src/main/java/com/marked/vifo/ui/activity/SignUpActivity.kt
package com.marked.vifo.ui.activity

import android.app.ProgressDialog
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.content.LocalBroadcastManager
import android.support.v7.app.AppCompatActivity
import com.marked.vifo.R
import com.marked.vifo.delegate.DialogRegistration
import com.marked.vifo.extra.GCMConstants
import com.marked.vifo.extra.ServerConstants
import com.marked.vifo.gcm.RegistrationBroadcastReceiver
import com.marked.vifo.gcm.service.LogInRegistrationIntentService
import com.marked.vifo.helper.add
import com.marked.vifo.helper.addToBackStack
import com.marked.vifo.ui.fragment.signup.SignUpEmailFragment
import com.marked.vifo.ui.fragment.signup.SignUpNameFragment
import com.marked.vifo.ui.fragment.signup.SignUpPassFragment
import org.jetbrains.anko.toast
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.net.SocketTimeoutException

class SignUpActivity : AppCompatActivity(), SignUpNameFragment.SignUpNameFragmentInteraction, SignUpEmailFragment.SignUpEmailFragmentInteraction, SignUpPassFragment.SignUpPassFragmentInteraction {
    private val mFragmentManager by lazy { supportFragmentManager }
    private val mProgressDialog by lazy { ProgressDialog(this) }
    private val mRegistrationBroadcastReceiver by lazy { RegistrationBroadcastReceiver(DialogRegistration(this, mProgressDialog)) }
    private val mBroadcastManagerastManager by lazy { LocalBroadcastManager.getInstance(this) }

    private var mName: String? = null
    private var mEmail: String? = null
    private var mPassword: String? = null

    override
    fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_up)

        mFragmentManager.add(R.id.fragmentContainer, SignUpNameFragment.newInstance())
    }

    override fun onResume() {
        super.onResume()
        mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, IntentFilter(GCMConstants.ACTION_SIGNUP))
        mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, IntentFilter(GCMConstants.ACTION_ERROR))
    }

    override
    fun onPause() {
        mBroadcastManagerastManager.unregisterReceiver(mRegistrationBroadcastReceiver)
        super.onPause()
    }

    override fun onSaveName(name: String) {
        mName = name
        mFragmentManager.addToBackStack(R.id.fragmentContainer, SignUpEmailFragment.newInstance(name))
    }

    override fun onSaveEmail(email: String) {
        mProgressDialog.show()
        mEmail = email
        checkEmail(mEmail)
    }

    override fun onSavePassword(password: String) {
        mPassword = password
        LogInRegistrationIntentService.startActionSignup(this, mName, mEmail, mPassword)
        mProgressDialog.setTitle("Signup")
        mProgressDialog.setMessage("Please wait...")
        mProgressDialog.isIndeterminate = true
        mProgressDialog.show()
    }

    /**
     * Sends a request to the server to check if the email is already in the database.
     * If that is true the user already has an account and we should tell him that
     * Else proceed to the next step
     * @param email the email adress to verify on the server
     */
    private fun checkEmail(email: String?) {
        val retrofit = Retrofit.Builder()
                .addConverterFactory(GsonConverterFactory.create())
                .baseUrl(ServerConstants.SERVER)
                .build()
        val service = retrofit.create(com.marked.vifo.networking.LoginAPI::class.java)
        service.hasAccount(email)
                .enqueue(object : Callback<Void> {
                    override fun onResponse(call: Call<Void>?, response: Response<Void>?) {
                        mProgressDialog.dismiss()
                        if (response?.isSuccess == true) {/* if hasAccount returns 200 -> user registered*/
                            toast("You already have an account")
                        } else /* go to next step*/
                            mFragmentManager.addToBackStack(R.id.fragmentContainer, SignUpPassFragment.newInstance())
                    }

                    override fun onFailure(call: Call<Void>?, error: Throwable?) {
                        mProgressDialog.dismiss()
                        when (error) {
                            is SocketTimeoutException -> toast("Timeout Error")
                        }
                        error?.printStackTrace()
                    }
                })
    }
}

